File: packages/react-native-nitro-modules/ios/core/RuntimeError.swift
//
//  RuntimeError.swift
//  NitroModules
//
//  Created by Marc Rousavy on 17.07.24.
//

/// Represents an error that occured at any point during the application's runtime.
///
/// Throw this error in Nitro Modules to provide clear and concise error messages to JS.
@frozen
public enum RuntimeError: Error, CustomStringConvertible {
  case error(withMessage: String)

  public var description: String {
    switch self {
    case .error(let message): return message
    }
  }

  /**
   * Creates a new `RuntimeError` from the given C++ `std::exception`.
   */
  public static func from(cppError: std.exception_ptr) -> RuntimeError {
    let message = margelo.nitro.getExceptionMessage(cppError)
    return .error(withMessage: String(message))
  }
}

extension Error {
  /**
   * Converts this `Error` to a C++ `std::exception`.
   */
  public func toCpp() -> std.exception_ptr {
    let message = String(describing: self)
    return margelo.nitro.makeException(std.string(message))
  }
}

File: packages/react-native-nitro-modules/ios/utils/RuntimeError.cpp
//
//  RuntimeError.cpp
//  NitroModules
//
//  Created by Marc Rousavy on 14.01.26.
//

#include "RuntimeError.hpp"
#include "NitroTypeInfo.hpp"
#include <NitroModules/Error.hpp>
#include <exception>
#include <string>

namespace margelo::nitro {

std::exception_ptr makeException(const std::string& message, const std::string& stacktrace) {
  return std::make_exception_ptr(nitro::Error(message, stacktrace));
}

std::string getExceptionMessage(const std::exception_ptr& exception) {
  if (exception == nullptr) [[unlikely]] {
    throw std::runtime_error("Cannot get error message of a nullptr exception_ptr!");
  }

  try {
    std::rethrow_exception(exception);
  } catch (const std::exception& error) {
    return error.what();
  } catch (...) {
    std::string errorName = TypeInfo::getCurrentExceptionName();
    return "Unknown " + errorName + " exception";
  }
}

std::string getExceptionStacktrace(const std::exception_ptr& exception) {
  if (exception == nullptr) [[unlikely]] {
    throw std::runtime_error("Cannot get error message of a nullptr exception_ptr!");
  }

  try {
    std::rethrow_exception(exception);
  } catch (const nitro::Error& error) {
    return error.stacktrace();
  } catch (...) {
    return "";
  }
}

} // namespace margelo::nitro
